presto-main/src/main/java/com/facebook/presto/operator/project/MergingPageOutput.java

```diff
@@ -0,0 +1,172 @@
+ * <p>
+ * Considering the CPU time required to process(filter, project) a full (~1MB) page returned by a
+ * connector, the CPU cost of memory copying (< 50kb, < 1024 rows) is supposed to be negligible
+ */
+@NotThreadSafe
+public class MergingPageOutput
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(ArrayBlockBuilder.class).instanceSize();
+
+    private final List<Type> types;
+    private final PageBuilder pageBuilder;
+    private final Queue<PageProcessorOutput> inputQueue = new LinkedList<>();
+    private final Queue<Page> outputQueue = new LinkedList<>();
+
+    private final long minPageSizeInBytes;
+    private final long minRowCount;
+
+    private boolean finishing = false;
+
+    public MergingPageOutput(Iterable<? extends Type> types, int minPageSizeInBytes, int minRowCount)
+    {
+        this(types, minPageSizeInBytes, minRowCount, DEFAULT_MAX_PAGE_SIZE_IN_BYTES);
+    }
+
+    public MergingPageOutput(Iterable<? extends Type> types, int minPageSizeInBytes, int minRowCount, int maxPageSizeInBytes)
+    {
+        this.types = ImmutableList.copyOf(requireNonNull(types, "types is null"));
+        this.minPageSizeInBytes = minPageSizeInBytes;
+        this.minRowCount = minRowCount;
+        pageBuilder = PageBuilder.withMaxPageSize(maxPageSizeInBytes, this.types);
+    }
+
+    public boolean needsInput()
+    {
+        return !finishing && outputQueue.isEmpty() && inputQueue.isEmpty();
+    }
+
+    public void addInput(PageProcessorOutput input)
+    {
+        requireNonNull(input, "input is null");
+        checkState(!finishing, "output is in finishing state");
+        inputQueue.add(input);
+    }
+
+    @Nullable
+    public Page getOutput()
+    {
+        if (!outputQueue.isEmpty()) {
+            return outputQueue.poll();
+        }
+
+        while (!inputQueue.isEmpty() && outputQueue.isEmpty()) {
+            Iterator<Page> input = inputQueue.peek();
+            if (input.hasNext()) {
+                process(input.next());
+            }
+            if (!input.hasNext()) {
+                inputQueue.poll();
+            }
+        }
+
+        if (inputQueue.isEmpty() && finishing) {
+            flush();
+        }
+
+        return outputQueue.poll();
+    }
+
+    public void finish()
+    {
+        finishing = true;
+    }
+
+    public boolean isFinished()
+    {
+        return finishing && inputQueue.isEmpty() && outputQueue.isEmpty() && pageBuilder.isEmpty();
+    }
+
+    private void process(Page page)
+    {
+        requireNonNull(page, "page is null");
+
+        // avoid memory copying for pages that are big enough
+        if (page.getSizeInBytes() >= minPageSizeInBytes || page.getPositionCount() >= minRowCount) {
+            flush();
+            outputQueue.add(page);
+            return;
+        }
+
+        buffer(page);
+    }
+
+    private void buffer(Page page)
+    {
+        for (int position = 0; position < page.getPositionCount(); position++) {
```

Comment:
Per @sopel39 sueggestion: replace inner and outer loop to make better use of CPU vectorization.